File: androidApp/src/main/java/com/habileducation/themovie/android/ui/movieDetail/MovieDetailViewModel.kt
package com.habileducation.themovie.android.ui.movieDetail

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.habileducation.themovie.viewModel.MovieDetailSharedViewModel
import com.habileducation.themovie.viewState.MovieDetailViewState
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.CoroutineDispatcher
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.collect
import kotlinx.coroutines.launch
import javax.inject.Inject

/**
 * Created by Annas Surdyanto on 18/11/21.
 *
 */

@HiltViewModel
class MovieDetailViewModel @Inject constructor(
    private val detailViewModel: MovieDetailSharedViewModel,
    private val dispatcher: CoroutineDispatcher
) : ViewModel() {
    val state = MutableStateFlow(MovieDetailViewState.empty)

    fun initState(movieId: Long) {
        viewModelScope.launch(dispatcher) {
            detailViewModel.fetchDetail(viewState = state.value, movieId).collect {
                state.value = it
            }
        }
    }

    fun setFavorite() {
        viewModelScope.launch(dispatcher) {
            detailViewModel.setFavorite(viewState = state.value).collect {
                state.value = it
            }
        }
    }
}

File: iosApp/iosApp/UI/MovieDetail/MovieDetailViewModel.swift
//
//  MovieDetailViewModel.swift
//  iosApp
//
//  Created by Annas Surdyanto on 29/11/21.
//  Copyright © 2021 private project. All rights reserved.
//

import shared

class MovieDetailViewModel : ObservableObject{
    private let sharedViewModel = MovieModule().movieDetailSharedViewModel
    
    @Published var state = MovieDetailViewState.companion.empty
    
    func initState(movieId: Int64){
        sharedViewModel.fetchDetail(viewState: state, movieId: movieId).collectCommon(){ [self] newState in
            self.state = newState!
        }
    }
    
    func setFavorite(){
        sharedViewModel.setFavorite(viewState: state).collectCommon(){ [self] newState in
            self.state = newState!
        }
    }
}
